### DISCUSSION TITLE
Should I add support for other httpx.Client arguments to TestClient?

### DISCUSSION THREAD

**Discussion by danielballan** [encode/starlette#1914] [Potential Issue] [answered]
Thanks to everyone who worked on moving TestClient to httpx in https://github.com/encode/starlette/pull/1376.

I notice it is missing some arguments including `timeout` and `event_hooks` and some others that may be useful to pass through to `httpx.Client.__init__`. I think it would be fairly straightforward to add these, and I would be happy to contribute a PR.

I can see two points that may need discussion:

* Exactly which parameters should be exposed? For example, should `verify` be omitted from the function signature, or should setting it anything but a default value raise?
* How should we handle default values? `httpx` uses sentinel objects that are not public API, so we could (1) define our own sentinel objects, (2) ask that `httpx` expose their default values as public API, (3) dig the sentinels objects out via signature inspection (seems overly clever to me but technically "public"), or (4) import the internal objects from httpx (seems bad but could be done).

If you would like I could contribute a PR with my best judgment and we could start the discussion from there, or we could first discuss here. Please let me know what would be most helpful. This would be my first contribution to the `encode` projects; my natural habitat is on the SciPy side.

**Comment by Kludex** [comment]
@tomchristie Do you have an opinion here?

**Reply by lovelydinosaur** [reply]
> Do you have an opinion here?

I've only got time for a quick comment just now, but to start with...

* `timeout` - Nope. timeouts on test clients just lets users write flaky tests.
* None of the SSL based parameters make sense in this context.

What else is omitted at the moment?

**Reply by danielballan** [reply]
I take your point on `timeout`, and yes I agree that all the SSL parameters make no sense in this context.

Omitted:

* `auth`
* `params`
* `headers`
* `http1`
* `http2`
* `follow_redirects`
* `max_redirects`
* `event_hooks`

Currently, these omissions can be worked around by setting properties on the `TestClient` object after construction, i.e.

```py
tc = TestClient()
tc.auth = ...
```

But it seems useful to accept them by `__init__` as well and pass them through to `httpx.Client`.

**Comment by Kludex** [comment, marked as answer]
Let's take each parameter as a separate subject, it will be easier to define "how it can be done now", and see how much are we saving to our users if we merge that on Starlette's TestClient itself.

**Reply by danielballan** [reply]
Thanks, sounds good.

Should I begin by starting a discussion for a given a parameter?

**Reply by Kludex** [reply]
That would be a good first step. 🙏
